server/src/main/java/org/elasticsearch/common/util/CollectionUtils.java

```diff
@@ -225,10 +227,26 @@ private int compare(byte[] left, byte[] right) {
         return ints.stream().mapToInt(s -> s).toArray();
     }
 
+    /**
+     * Deeply inspects a Map, Iterable, or Object array looking for references back to itself.
+     * @throws IllegalArgumentException if a self-reference is found
+     * @param value The object to evaluate looking for self references
+     */
     public static void ensureNoSelfReferences(Object value) {
```

Comment:
I think we should make the hit mandatory and remove this method.

Reply:
Nice -- I always like less ambiguity in error situation especially for testing. I can throw that change into this PR if that works for you.